### ISSUE TITLE
Cloner doesn't handle subList properly, results in ConcurrentModificationException

### ISSUE THREAD

**Issue by emichael** [kostaskougios/cloning#72]
```
List<String> a = new ArrayList<>();
a.add("1");
a.add("2");
a.add("3");

List<String> b = a.subList(0, 1);

b.add("3");
b.remove(0);

b.size(); // fine

new Cloner().deepClone(b).size(); // throws ConcurrentModificationException
```

The issue is that the subList and the underlying array must have the same modCount. However, the Cloner uses the ArrayList fast cloner for the underlying array, which changes the modCount in the clone. The modCount of the cloned subList stays the same.

I'm not sure of the best way to fix this. Maybe reflection should be used to copy the old modCount to the new ArrayList?

**Comment by emichael** [comment]
Just for context, I'm using this library as a part of a larger framework that emulates a distributed message-passing environment. The cloning library clones messages. I would like to make the cloning as transparent to the users of this framework as possible.

One reasonable approach for me would be to throw a NotSerializableException when a non-Serializable object is cloned by overriding `cloneInternal` (trying to emulate the behavior of cloning with standard Java serialization). This would work because the Java Collections library is careful about not marking "derived" data structures like `List.subList(...)` and `Map.keySet()` Serializable. The only issue there is that many objects I would like to clone have transient non-Serializable fields. I thought about trying to only throw the NotSerializableException for non-transient non-Serializable fields (thereby throwing an exception for an object iff the standard Java serialization would throw an exception). It doesn't look like there's a clean way to do this with the `Cloner` as-is, though. Any advice?

**Comment by kostaskougios** [comment]
Actually the issue was a bit weird but I've pushed a workaround , artifact 1.9.10-SNAPSHOT. Please let me know if it resolves the issue.

So ArrayList.subList returns an instance of the private ArrayList.SubList class. The cloner didn't use the FastClonerArrayList because the instance was not ArrayList. SubList now is an inner class of ArrayList hence it has a pointer to it's parent class. But seems when it was cloned, the parent class was cloned with the FastCloner and as you said probably some variables were out of sync with what SubList expected.

I've added a fast cloner for SubList and your sample code works.

**Comment by emichael** [comment]
Thanks for the quick fix! Unfortunately, `ArrayList$SubList` isn't the only `SubList`. You could see the same issue with the `AbstractList$SubList` (and potentially `AbstractList$RandomAccessSubList`) if the backing list is a `LinkedList`, for example. Looking through the other classes which are fast cloned, it looks like the `keySet()` etc. in `HashMap` should be fine? Not sure about some of the others. Iterators and spliterators will definitely be problematic for the same reason, but I don't know how much effort they're worth, or what could even be done (after all, who would want to clone those?).

**Comment by kostaskougios** [comment]
For LinkedList I added a fast cloner for both sublists. But it becomes more of a problem with iterator and the modCount access subclasses do to their fast-cloned parent classes. I'll try to think of something but in the meantime as you said fastcloners can be disabled

**Comment by emichael** [comment]
Thanks for handling all of this! Will you be releasing a new version soon?

**Comment by kostaskougios** [comment]
Well, I can release the List fixes if it helps.

**Comment by emichael** [comment]
That's what I meant. It would be helpful when you get a chance.

**Comment by kostaskougios** [comment]
np, released 1.9.10

**Comment by emichael** [comment]
Thanks again! This library has been really useful.
